### PR TITLE
artiq_flash: Add --no-variant flag to just use given path verbatim

### PR DESCRIPTION
**Pull request by dnadlinger**
@sbourdeauducq: We have been using a local patch like this for a long, long time, as our Buildbot setup just archives gateware builds into a uniform directory structure by (Buildbot) build target rather than bothering with variants – okay to merge?

### CODE REVIEW

File: artiq/frontend/artiq_flash.py [hunk 1]
```diff
@@ -346,6 +348,9 @@ def main():
         action in args.action
         for action in ["gateware", "rtm_gateware", "bootloader", "firmware", "load", "rtm_load"])
     variant = args.variant
+    if variant is None and args.no_variant:
+        # Just use default actions in the below.
+        variant = ""
     if needs_artifacts and variant is None:
```
Comment (sbourdeauducq) [inline artiq/frontend/artiq_flash.py]: This (variant listing) still gets executed when ``--no-variant`` is given. Is that OK?

**Comment by sbourdeauducq** [comment]
The ``artiq_flash`` variant/directory handling is already quite hairy as it is. Is it really the best solution to add more complexity and corner cases to it? Why is your build system not using variants?

File: artiq/frontend/artiq_flash.py (hunk 1 above)
Comment (dnadlinger) [reply artiq/frontend/artiq_flash.py]: Is it? Variant should no longer be `None` then.

Give me ~1h to sort out the SiPyCo deployment so I can test this in situ…

**Comment by dnadlinger** [comment]
You'd need to ask @cjbe about the details, but I suspect he would say that the answer to 

> The artiq_flash variant/directory handling is already quite hairy as it is. Is it really the best solution to add more complexity and corner cases to it? 

was precisely just not to use it.

Also, from a practical perspective during development, not being forced into the variant directory structure is really handy. For example, while iterating on firmware, I tend to just `python -m artiq.gateware.targets … --no-compile-gateware` and `artiq_flash … -d <build_dir> --no-variant firmware start`, without having to copy files around to make artiq_flash happy.

From a design point of view, handling all the OpenOCD setup/target-specific memory ranges/… seems like neatly independent piece of functionality; the variant directory lookup stuff being a pretty orthogonal concern.

**Comment by sbourdeauducq** [comment]
> For example, while iterating on firmware, I tend to just `python -m artiq.gateware.targets … --no-compile-gateware` and `artiq_flash … -d <build_dir> --no-variant firmware start`, without having to copy files around to make artiq_flash happy.

This is what ``--srcbuild`` is for.

File: artiq/frontend/artiq_flash.py (hunk 1 above)
Comment (dnadlinger) [reply artiq/frontend/artiq_flash.py]: Works as intended; the variant listing is not executed.

**Comment by dnadlinger** [comment]
> This is what --srcbuild is for.

Well, `--no-variant` works just the same, except that it's more flexible.

Just for context, we have a central archive of several hundred builds at this point, with some twenty targets each.

This is integrated into our tooling, such that with a simple local configuration mapping logical names to USB hostnames and Kasli serial numbers, we can do things like `./flash.py -n 1552` to flash all DRTIO nodes to a certain build number, or e.g. `./flash -d trap_satellite -g bcd86` to flash just one node to a specific Git hash.

All of this works beautifully and makes it easy even for barely-programmers to iterate on gateware configs or firmware changes. We just need to be able to use the target-related functionality of `artiq_flash` without it also enforcing unrelated notions of what the directory layout is supposed to be. (The latter was only added fairly recently in ARTIQ 4/ 61c64a76bea0aece490daa89f43b93f07e820c9a.)

**Comment by sbourdeauducq** [comment]
> Well, `--no-variant` works just the same, except that it's more flexible.

Would it support flashing Sayma (both AMC and RTM at the same time) from a source build?

**Comment by dnadlinger** [comment]
> Would it support flashing Sayma (both AMC and RTM at the same time) from a source build?

I don't particularly care about the merits of `--srcbuild`; I just want to be able to use the pre-ARTIQ-4 behaviour of artiq_flash.

**Comment by sbourdeauducq** [comment]
> I don't particularly care about the merits of `--srcbuild`; I just want to be able to use the pre-ARTIQ-4 behaviour of artiq_flash.

I was asking because, short of a nice complete refactoring, I am trying to reduce the complexity and amount of corner cases. Removing ``--srcbuild`` and replacing it with ``--no-variant`` would have helped achieve that goal - but we need to flash Sayma.

**Comment by sbourdeauducq** [comment]
``artiq_flash`` will be simplified with a reworking of the bitstream build service.

**Pull request closed without merging**